### PR TITLE
bpo-46033: Clarify for-statement execution

### PR DESCRIPTION
**Pull request by michcio1234**
In for statement description, there seem to be two sentences meant to mean the same:

> The suite is then executed once for each item provided by the iterator, in the order returned by the iterator. Each item in turn is assigned to the target list using the standard rules for assignments (see Assignment statements), and then the suite is executed.

(from https://docs.python.org/3/reference/compound_stmts.html#the-for-statement)

I modified this text to be - in my opinion - cleaner.

If I am wrong, and the current version is actually correct, then it is unclear - to me it sounds like the iterator is iterated through twice, and suite is executed twice for each item.

https://bugs.python.org/issue46033

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@michcio1234

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by terryjreedy** [comment]
Please let us know when you have signed the CLA.  It might take a few days to be recorded.

**Comment by michcio1234** [comment]
@terryjreedy done. Forgot to check myself.

**Comment by michcio1234** [comment]
@terryjreedy added news entry to make Azure Pipelines PR check green. CLA is also signed.

**Comment by JelleZijlstra** [comment]
@michcio1234 there is a merge conflict now, could you take a look?

Also, I don't think this change needs a NEWS entry. I added the "skip news" label so CI won't fail if you remove the NEWS entry you added.

**Review by terryjreedy — DISMISSED** [review]
If the change was to just remove a sentence, then no news would be needed.  But we did something else, and I made more edits to blend with the other, conflicting merge.  Hence I changed both title and news.

I think this is ready, but please read to check.

Note: the other addition was not backported, so this cannot be either.

File: Doc/reference/compound_stmts.rst [hunk 1]
```diff
@@ -157,17 +157,16 @@ The :keyword:`for` statement is used to iterate over the elements of a sequence
...
-list can contain starred elements (``*x, *y``) that will be unpacked in the
-final iterator (as when constructing a ``tuple`` or ``list`` literal). The
-suite is then executed once for each item provided by the iterator, in the
-order returned by the iterator.  Each item in turn is assigned to the target
-list using the standard rules for assignments (see :ref:`assignment`), and then
-the suite is executed.  When the items are exhausted (which is immediately when
-the sequence is empty or an iterator raises a :exc:`StopIteration` exception),
-the suite in the :keyword:`!else` clause, if present, is executed, and the loop
-terminates.
+The starred list expression is evaluated once; it should yield an iterable object.
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
The ``starred_list`` expression is evaluated once; it should yield an iterable object.
```
Make it clear that we're referring back to the grammar production

File: Doc/reference/compound_stmts.rst [hunk 2]
```diff
@@ -157,17 +157,16 @@ The :keyword:`for` statement is used to iterate over the elements of a sequence
...
+If the expression is a sequence, it can contain starred elements (``*x, *y``)
+that will be unpacked, as when constructing a ``tuple`` or ``list`` literal.
+An iterator is created for the resulting iterable.  The first item provided
+by the iterator is then assigned to the target list using the standard
+rules for assignments (see :ref:`assignment`), and the suite is executed - this
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
rules for assignments (see :ref:`assignment`), and the suite is executed. This
```

File: Doc/reference/compound_stmts.rst [hunk 3]
```diff
@@ -157,17 +157,16 @@ The :keyword:`for` statement is used to iterate over the elements of a sequence
...
+repeats for each item provided by the iterator.  When the items are exhausted
+(which is immediately when the sequence is empty or an iterator raises
+a :exc:`StopIteration` exception), the suite in the :keyword:`!else` clause,
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]:
```suggestion
repeats for each item provided by the iterator.  When the iterator is exhausted,
the suite in the :keyword:`!else` clause,
```
We don't have to explain here how iterators work, and the current text is imprecise for classes using the old-style iteration protocol.

Perhaps we can link to the `iterator` glossary entry though.

File: Doc/reference/compound_stmts.rst [hunk 4]
```diff
@@ -157,17 +157,16 @@ The :keyword:`for` statement is used to iterate over the elements of a sequence
...
-suite is then executed once for each item provided by the iterator, in the
-order returned by the iterator.  Each item in turn is assigned to the target
-list using the standard rules for assignments (see :ref:`assignment`), and then
-the suite is executed.  When the items are exhausted (which is immediately when
-the sequence is empty or an iterator raises a :exc:`StopIteration` exception),
-the suite in the :keyword:`!else` clause, if present, is executed, and the loop
-terminates.
+The starred list expression is evaluated once; it should yield an iterable object.
+If the expression is a sequence, it can contain starred elements (``*x, *y``)
+that will be unpacked, as when constructing a ``tuple`` or ``list`` literal.
```
Comment (JelleZijlstra) [inline Doc/reference/compound_stmts.rst]: Do we need this sentence? It's not part of the semantics of the for loop itself, but of the `starred_list` expression.

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (terryjreedy) [reply Doc/reference/compound_stmts.rst]: As I remember from experiments a few years ago, old-style iteration is done via a new-style iterator that serves as an adaptor.  The looping code only sees the adaptor and does not know that it wraps a get_item that raises IndexError.  But I agree that we do not need to explain 'exhausted' again here.  Let's link the first uses of both `iterable` and `iterator`.

**Comment by terryjreedy** [comment]
I will re-approve with 2 links added (if you agree) and removal of sentence as suggested.

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (JelleZijlstra) [reply Doc/reference/compound_stmts.rst]: Agree, cross-references to the definitions of `iterable` and `iterator` would be good.

File: Doc/reference/compound_stmts.rst (hunk 4 above)
Comment (terryjreedy) [reply Doc/reference/compound_stmts.rst]: gone

File: Doc/reference/compound_stmts.rst (hunk 3 above)
Comment (terryjreedy) [reply Doc/reference/compound_stmts.rst]: done

**Comment by terryjreedy** [comment]
Will merge after checking generated .html on my machine.

**Comment by michcio1234** [comment]
@terryjreedy @JelleZijlstra thank you very much for taking care of this and sorry for not responding lately. Glad that something useful came out of this. :)

**Pull request merged**